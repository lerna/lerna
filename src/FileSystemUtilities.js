import pathExists from "path-exists";
import logger from "./logger";
import mkdirp from "mkdirp";
import rimraf from "rimraf";
import fs from "fs";

var ENDS_WITH_NEW_LINE = /\n$/;

function ensureEndsWithNewLine(string) {
  return ENDS_WITH_NEW_LINE.test(string) ? string : string + "\n";
}

export default class FileSystemUtilities {
  @logger.logifySync
  static mkdirSync(filePath) {
    fs.mkdirSync(filePath);
  }

  @logger.logifyAsync
  static mkdirp(filePath, callback) {
    mkdirp(filePath, callback);
  }

  @logger.logifyAsync
  static readdirSync(filePath) {
    return fs.readdirSync(filePath);
  }

  @logger.logifySync
  static existsSync(filePath) {
    return pathExists.sync(filePath);
  }

  @logger.logifyAsync
  static writeFile(filePath, fileContents, callback) {
    fs.writeFile(filePath, ensureEndsWithNewLine(fileContents), callback);
  }

  @logger.logifySync
  static writeFileSync(filePath, fileContents) {
    fs.writeFileSync(filePath, ensureEndsWithNewLine(fileContents));
  }

  @logger.logifySync
  static readFileSync(filePath) {
    return fs.readFileSync(filePath, "utf-8").toString().trim();
  }

  @logger.logifyAsync
  static rimraf(filePath, callback) {
    rimraf(filePath, callback);
  }

  @logger.logifySync
  static unlinkSync(filePath) {
    fs.unlink(filePath);
  }
}
