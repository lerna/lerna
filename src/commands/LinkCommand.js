import Command from "../Command";
import PackageUtilities from "../PackageUtilities";

export function handler(argv) {
  return new LinkCommand(argv._, argv).run();
}

export const command = "link";

export const describe = "Symlink together all packages which are dependencies of each other";

export const builder = {};

export default class LinkCommand extends Command {
  get defaultOptions() {
    return {};
  }

  initialize(callback) {
    this.exact = this.options.exact;

    callback(null, true);
  }

  execute(callback) {
    const {packages, packageGraph, logger} = this;

    PackageUtilities.symlinkPackages(packages, packageGraph, logger, callback);
  }
}
