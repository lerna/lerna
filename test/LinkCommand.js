import log from "npmlog";

// mocked or stubbed modules
import FileSystemUtilities from "../src/FileSystemUtilities";

// helpers
import callsBack from "./helpers/callsBack";
import exitWithCode from "./helpers/exitWithCode";
import initFixture from "./helpers/initFixture";
import normalizeRelativeDir from "./helpers/normalizeRelativeDir";

// file under test
import LinkCommand from "../src/commands/LinkCommand";

// silence logs
log.level = "silent";

// stub symlink in certain tests to reduce redundancy
const fsSymlink = FileSystemUtilities.symlink;
const resetSymlink = () => {
  FileSystemUtilities.symlink = fsSymlink;
};
const stubSymlink = () => {
  FileSystemUtilities.symlink = jest.fn(callsBack());
};

// object snapshots have sorted keys
const symlinkedDirectories = (testDir) =>
  FileSystemUtilities.symlink.mock.calls.map((args) => {
    return {
      _src: normalizeRelativeDir(testDir, args[0]),
      dest: normalizeRelativeDir(testDir, args[1]),
      type: args[2],
    };
  });

describe("LinkCommand", () => {
  beforeEach(() => {
  });

  afterEach(() => jest.resetAllMocks());

  describe("with local package dependencies", () => {
    let testDir;

    beforeEach(() => initFixture("LinkCommand/basic").then((dir) => {
      testDir = dir;
    }));

    beforeEach(stubSymlink);
    afterEach(resetSymlink);

    it("should symlink all packages", (done) => {
      const linkCommand = new LinkCommand([], {}, testDir);

      linkCommand.runValidations();
      linkCommand.runPreparations();

      linkCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);

        try {
          expect(symlinkedDirectories(testDir)).toMatchSnapshot();

          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });
  });
});
