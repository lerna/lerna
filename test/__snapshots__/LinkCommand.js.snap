// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkCommand with local package dependencies should symlink all packages 1`] = `
Array [
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-2/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-3/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2",
    "dest": "packages/package-3/node_modules/@test/package-2",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2/cli.js",
    "dest": "packages/package-3/node_modules/.bin/package-2",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3",
    "dest": "packages/package-4/node_modules/package-3",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-3/cli1.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli1",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3/cli2.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli2",
    "type": "exec",
  },
]
`;
